enhancements/network/nodeport-range-configuration.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Make Service NodePort Range Configurable
+authors:
+  - "@abhat"
+reviewers:
+  - "@dcbw"
+  - "@deads2k"
+  - "@smarterclayton"
+  - "@danwinship"
+  - "@knobunc"
+approvers:
+  - "@knobunc"
+creation-date: 2020-07-06
+last-updated: 2020-07-06
+status: implementable
+---
+# Configurability of Service NodePort Range
+This enhancement targets making the service node-port range of an Openshift cluster configurable.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Service node-port range is a [configurable parameter](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/v4.1.0/config/defaultconfig.yaml#L147)
+on the Kubernetes API server. As our customers scale the number of node-port services in their Openshift clusters,
+there is a desire to make this node port range configurable.
+
+## Motivation
+
+Service node-port range is a configurable parameter consumed by the Kubernetes API server as
+a slice of port ranges. It defaults currently to a single port range (30000-32767).
+It is consumed in Openshift via the ConfigMap resource that holds all the api-server configuration.
+In order to allow customers to scale the number of node-port services in the cluster to a value greater
+than that allowed by the default service node-port range, it is important to allow dynamically changing this
+range to allow for more ports. This enhancement focuses on the mechanics of allowing users to configure a
+custom service node-port range for their clusters.
+
+### Goals
+
+- Customers can configure the service node-port range for their cluster during install.
+- Customers can configure the service node-port range for their cluster post install as their needs change.
+
+### Non-Goals
+
+- Services launched prior to customers changing the service node-port range will not work if the new range is
+  not inclusive of the old range.
+- This enhancement makes no guarantees of providing non-disruptive change to the service node-port range.
```

Comment:
Why was this chosen?  What's the justification?

Reply:
Based on the range(s) the users configures, the ports that were previously opened for the infrastructure may not necessarily stay open after a change. So there is no guarantee that (unless the new range is inclusive of the old range), that the services will continue to work. If the range is however inclusive, we can guarantee that older services will continue to work. I have validated this case using the PRs mentioned in this enhancement.

Reply:
Can you give an example of a component that would break?  I.e. I have service X that has node port 33001. We change the range to stop at 33000.  Which component breaks?  Programming firewalls?  Service load balancer?  Kube-proxy?  OVN?

Reply:
Service X would break; people would (presumably) still be trying to reach it on port 33001, but the underlying network (cloud security groups etc) would no longer allow that traffic.

Reply:
... I thought it will eventually be deleted by the portallocator report loop but seems it will record an event only
https://github.com/kubernetes/kubernetes/blob/1c11ff7a26c498dc623f060aa30c7c970f3d3eee/pkg/registry/core/service/portallocator/controller/repair.go#L154

Reply:
The other option would be to forbid a shrink if there were still services that fell outside the new range.

Reply:
Which implies the cluster-kube-apiserver-operator needs to monitor services to check if any of them is of nodePort type and the current node port allocation for any service falls outside of the new range? Sounds doable, but not a clean separation of concerns.

Reply:
We'd only need to check at startup, but yeah... it's ugly since they are separate processes.  Can we at least scan and alert if we see it?

What does the kubeproxy do when the range has shrunk and there is still a service with an ExternalIP outside the range?

Reply:
the nodePort range is enforced in the API, when you create a Service with a NodePort, the apiserver checks the portallocator (a bitmap stored in etcd) and allocates the port to that service if available.

Kubeproxy reads the service object, so it should not be affected if the service was already created ... 

checking the kubernetes code, it also seems that any update on this service with the NodePort out of range will fail, because it won't be able to allocate the port out of range.
My reading is that the service will keep working, until it is deleted or updated with a new node port within the new range
However, better double check this theory :smile:

Reply:
Yes 👍 to what @aojea said. kubeproxy reads the service object, it is not affected if the service was already created. I actually verified this in my dev-testing. Existing services continue to work if I give a range disjoint from the previous setting so long as the security groups underneath are untouched or the changes to them are additive.